spec.html

```diff
@@ -824,10 +824,7 @@ <h1>Implicit Completion Values</h1>
           1. Assert: _completionRecord_ is a Completion Record.
           1. Return _completionRecord_ as the Completion Record of this abstract operation.
         </emu-alg>
-        <p>A &ldquo;return&rdquo; statement without a value in an algorithm step means the same thing as:</p>
-        <emu-alg>
-          1. Return NormalCompletion(*undefined*).
-        </emu-alg>
+        <p>If all steps in an abstract operation have been executed without returning, the abstract operation behaves as though there were an additional step reading "Return NormalCompletion(~unused~)." after the existing steps. ~unused~ is a special value which is never directly consumed.</p>
```

Comment:
"If all steps in an abstract operation have been executed without returning..."

Well, not necessarily *all* steps, because if there are If-else-steps, we might have executed only one of the arms. Maybe "... completes execution without returning ..." ?

Also, it's not just abstract operations, there's at least one syntax-directed operation (ForDeclarationBindingInstantiation) that this applies to. (Or maybe you consider SDOs a subset of AOs, which the spec isn't entirely clear on.)

---

"`~unused~` is a special value which is never directly consumed."

So it might be *indirectly* consumed? What does that even mean?

Maybe "`~unused~` is an ad hoc value that is only used for this purpose."

---

Also maybe "Operations that return `~unused~` are typically invoked in a `Perform` step." (I only found one exception, where `AsyncGeneratorEnqueue` has `Let _check_ be AsyncGeneratorValidate(...).`)

Reply:
I'd say "Operations that return `~unused~` must only be invoked in a `Perform` step.`

Reply:
So then what about the `AsyncGeneratorEnqueue` case?

Reply:
@jmdyck

> Well, not necessarily all steps, because if there are If-else-steps, we might have executed only one of the arms.

I was modeling the if-else as a single step (with the arms being substeps), but I can see how your interpretation is plausible.

> Maybe "... completes execution without returning ..." ?

I'd be OK with that, though I don't love it. Open to other phrasings here.

I originally said something like "if control reaches the end of the algorithm", but I felt that was maybe too confusing, since we normally talk about control-flow of the JS being executed rather than of spec steps.

> Also, it's not just abstract operations, there's at least one syntax-directed operation (ForDeclarationBindingInstantiation) that this applies to. (Or maybe you consider SDOs a subset of AOs, which the spec isn't entirely clear on.)

For avoidance of doubt I'll list SDOs explicitly.

> So it might be _indirectly_ consumed? What does that even mean?

It means the completion containing it is consumed, or it's stored in a binding within a macro but the binding is not used after the macro, etc. I am content with this phrasing personally; does anyone else want to weigh in?

@ljharb 
> Operations that return `~unused~` must only be invoked in a `Perform` step

That's not a guarantee I want to make: it is reasonable to have instances of the pattern

```
1. Let _result_ be SomeProcedure(whatever).
1. If _result_ is an abrupt completion, then
  1. Do some cleanup. 
```

Reply:
Hmm, that's a good question. I don't have a suggestion.

Reply:
> I was modeling the if-else as a single step (with the arms being substeps),

Indeed, that's how structured programming wants us to think of it, but when each `Else` has a different step-number, it's harder to suggest that they're not different steps. 

> > Maybe "... completes execution without returning ..." ?
> 
> I'd be OK with that, though I don't love it.

Yeah, I'm not thrilled with it either. The problem is, the existing spec gives us almost no terminology/model for the execution of algorithms. 

> I originally said something like "if control reaches the end of the algorithm", but I felt that was maybe too confusing, since we normally talk about control-flow of the JS being executed rather than of spec steps.

Yeah, I don't see any examples of the spec using "control" in the context of performing spec steps. At least for "execute" there are a handful of examples.